examples/vision/vivit.py

```diff
@@ -0,0 +1,444 @@
+        test_videos = data["test_images"]
+
+        # Get labels
+        train_labels = data["train_labels"].flatten()
+        valid_labels = data["val_labels"].flatten()
+        test_labels = data["test_labels"].flatten()
+
+    return (
+        (train_videos, train_labels),
+        (valid_videos, valid_labels),
+        (test_videos, test_labels),
+    )
+
+
+# Get the metadata of the dataset
+info = INFO[DATASET_NAME]
+
+# Get the dataset
+prepared_dataset = download_and_prepare_dataset(info)
+(train_videos, train_labels) = prepared_dataset[0]
+(valid_videos, valid_labels) = prepared_dataset[1]
+(test_videos, test_labels) = prepared_dataset[2]
+
+"""
+### `tf.data` pipeline
+"""
+
+
+@tf.function
+def preprocess(frames: tf.Tensor, label: tf.Tensor):
+    """
+    Preprocess the frames tensors and parse the labels
+    """
+    # Preprocess images
+    frames = tf.image.convert_image_dtype(
+        frames[
+            ..., tf.newaxis
+        ],  # the new axis is to help for further processing with Conv3D layers
+        tf.float32,
+    )
+
+    # Parse label
+    label = tf.cast(label, tf.float32)
+    return frames, label
+
+
+def prepare_dataloader(
+    videos: np.ndarray,
+    labels: np.ndarray,
+    loader_type: str = "train",
+    batch_size: int = BATCH_SIZE,
+):
+    """
+    Utility function to prepare dataloader.
+    """
+    dataset = tf.data.Dataset.from_tensor_slices((videos, labels))
+
+    if loader_type == "train":
+        dataset = dataset.shuffle(1024)
+
+    dataloader = (
+        dataset.map(preprocess, num_parallel_calls=tf.data.AUTOTUNE)
+        .batch(batch_size)
+        .prefetch(tf.data.AUTOTUNE)
+    )
+
+    return dataloader
+
+
+trainloader = prepare_dataloader(train_videos, train_labels, "train")
+validloader = prepare_dataloader(valid_videos, valid_labels, "valid")
+testloader = prepare_dataloader(test_videos, test_labels, "test")
+
+"""
+## Tubelet Embedding
+
+In ViTs an image is divided into patches which is then spatially
+flattened and projected as a tokenization scheme. For a video one can
+repeat this process for individual frames. **Uniform frame sampling**
+as suggested by the authors is a tokenization scheme in which we
+sample frames from the video clip and perform simple ViT tokenization.
+
+| ![uniform frame sampling](https://i.imgur.com/aaPyLPX.png) |
+| :--: |
+| Uniform Frame Sampling [Source](https://arxiv.org/abs/2103.15691) |
+
+**Tubelet Embedding** is different in terms of capturing the temporal
+information. From the video we extract volumes. These volumes contain
+patches of the frame and the temporal information as well. The volumes
+are then flattened and projected to build video tokens.
+
+| ![tubelet embedding](https://i.imgur.com/9G7QTfV.png) |
+| :--: |
+| Tubelet Embedding [Source](https://arxiv.org/abs/2103.15691) |
+"""
+
+
+class TubeletEmbedding(layers.Layer):
+    def __init__(self, embed_dim=PROJECTION_DIM, patch_size=PATCH_SIZE):
```

Comment:
Make these positional args (no default) and pass the hardcoded value upon instantiation -- this makes the layer easier to reuse elsewhere